qiskit/compiler/assembler.py

```diff
@@ -34,13 +34,47 @@ def assemble_circuits(circuits, qobj_id=None, qobj_header=None, run_config=None)
         qobj_header (QobjHeader): header to pass to the results
         run_config (RunConfig): configuration of the runtime environment
 
+    Raises:
+        QiskitError: if run_config parameters are not compatible with circuit parameters
+
     Returns:
         QasmQobj: the Qobj to be run on the backends
     """
     qobj_config = QasmQobjConfig()
     if run_config:
         qobj_config = QasmQobjConfig(**run_config.to_dict())
 
+    # Bind parameters (if specified) or circuits are parameterized
+    parameter_binds = getattr(run_config, 'parameter_binds', [])
+    if parameter_binds or \
+       any(circuit.parameters for circuit in circuits):
+
+        all_bind_parameters = [bind.keys()
+                               for bind in parameter_binds]  # pylint: disable=not-an-iterable
+        all_circuit_parameters = [circuit.parameters for circuit in circuits]
+
+        # Collect set of all unique parameters across all circuits and binds
+        unique_parameters = set(param
+                                for param_list in all_bind_parameters + all_circuit_parameters
+                                for param in param_list)
+
+        # Check that all parameters are common to all circuits and binds
+        if not all_bind_parameters \
+           or not all_circuit_parameters \
+           or any(unique_parameters != bind_params for bind_params in all_bind_parameters) \
+           or any(unique_parameters != parameters for parameters in all_circuit_parameters):
+            raise QiskitError(
+                ('Mismatch between run_config.parameter_binds and all circuit parameters. ' +
+                 'Parameter binds: {} ' +
+                 'Circuit parameters: {}').format(all_bind_parameters, all_circuit_parameters))
```

Comment:
Couldn't we move all the checks from line 49 to 69 inside `_parse_run_args`, and keep the `assemble_circuits` function lean? Basically the `assemble_circuits` function assumes that it receives a valid `run_config`, and that all checks have been done.

Reply:
Sort of, but the verification here doesn't just depend on the args, it also depends on the parameters present in the circuit. I pulled both the verification and expansion out into a private function that's run by `assemble` before `assemble circuits` in 77925cea87dacf62354d06c0bbf6a7c20ef9f59c. Thoughts?

Reply:
Sounds good.